### PR TITLE
SI-8237 Avoid cyclic constraints when inferring hk type args

### PR DESCRIPTION
**Pull request by retronym**
An `AppliedTypeVars` spawned from `HKTypeVar#applyArgs`
(necessarily) shares the same `TypeConstraints`.

But, we can have multiple ATVs based on a single HKTV floating
around during inference, and they can appear on both sides
of type relations. An example of this is traced out in
the enclosed test.

This commit avoids registering upper/lower bound constraints
when this is detected.

In the enclosed test, we end up with an empty set of constraints
for `?E`, which results in inference of Nothing, which is what
we expect.

I have also improved the printing of ATVs (to include the args)
and sharpened the log message when `solve` leaves type variables
instantiated to `NoType`, rather than some other type that doesn't
conform to the bounds. Both of these changes helped me to get
to the bottom of this ticket.

The reported test has quite a checkered history:
- in 2.10.0 it worked, but more by good luck than good planning
- after the fix for SI-7226 / 221f52757aa6, it started crashing
- from 3bd897ba0054f (a merge from 2.10.x just before 2.11.0-M1)
  we started getting a type inference failure, rather than a crash.
  "no type parameters for method exists [...] because cyclic
  instantiation".
- It still crashes in `isGround` in 2.10.3.

Review by @adriaanm

### CODE REVIEW

**Comment by xeno-by** [comment]
@retronym It's a genuine failure:

```
BUILD FAILED
/localhome/jenkins/a/workspace/pr-scala-rangepos/scala/build.xml:1434: The following error occurred while executing this line:
/localhome/jenkins/a/workspace/pr-scala-rangepos/scala/build-ant-macros.xml:508: Test suite finished with 1 case failing:
fail - run/t7319.scala  [output differs]% scalac t7319.scala
...
% diff /localhome/jenkins/a/workspace/pr-scala-rangepos/scala/test/files/run/t7319-run.log /localhome/jenkins/a/workspace/pr-scala-rangepos/scala/test/files/run/t7319.check
@@ -21,7 +21,7 @@ scala> convert(Some[Int](0))
  --- because ---
 argument expression's type is not compatible with formal parameter type;
  found   : Some[Int]
- required: ?F(_$1) forSome { type _$1 <: ?F(_$2) forSome { type _$2 } }
+ required: ?F forSome { type _$1 <: ?F forSome { type _$2 } }
               convert(Some[Int](0))
               ^
 <console>:12: error: type mismatch;
```

File: src/reflect/scala/reflect/internal/Types.scala [hunk 1]
```diff
@@ -3045,6 +3051,17 @@ trait Types
       case ConstantTrue  => true
       case tv: TypeVar   => tv.suspended
     }
+
+    /**
+     * `AppliedTypeVar`s share the same `TypeConstraint` with the `HKTypeVar` that it was spawned from.
+     * A type inference session can also have more than one ATV.
+     * If we don't detect that, we end up with "cyclic contraint" when we try to instatiate type parameters
```
Comment (xeno-by) [inline src/reflect/scala/reflect/internal/Types.scala]: conStraint

Comment (xeno-by) [inline src/reflect/scala/reflect/internal/Types.scala]: instaNtiate

Comment (som-snytt) [inline src/reflect/scala/reflect/internal/Types.scala]: One of the lint tools they're going to focus on in 2014 is spellcheck.

You can either have retronym answering email questions, stack overflow, code review, keeping paulp in the loop, maybe writing some code, or you can have Queen's English.

Maybe we can help by standardizing, e.g., interpol8. Or `i12` for the noun.

More devastating is the nonstandard doc comment, which should have `/** App...` etc.

**Comment by retronym** [comment]
That just needed a checkfile update (`AppliedTypeVar#toString` now prints its type arguments, this improves error messages after inference failure.)

And I've remedied the abominations against the Queen's English and Scaladocumentarians.

**Comment by adriaanm** [comment]
LGTM modulo benign checkfile update (braces have gotten a bit more square)

**Comment by retronym** [comment]
`git diff 0cee06b..1cf1bae`

``` diff
diff --git a/test/files/neg/names-defaults-neg.check b/test/files/neg/names-defaults-neg.check
index ad3ac60..20ddd55 100644
--- a/test/files/neg/names-defaults-neg.check
+++ b/test/files/neg/names-defaults-neg.check
@@ -88,7 +88,7 @@ names-defaults-neg.scala:76: error: no type parameters for method test4: (x: T[T
  --- because ---
 argument expression's type is not compatible with formal parameter type;
  found   : List[Int]
- required: ?T(?T(List[?T(X forSome { type X })]))
+ required: ?T[?T[List[?T[X forSome { type X }]]]]
 Error occurred in an application involving default arguments.
   test4()
   ^
```

**Comment by adriaanm** [comment]
Also, thanks for improving the logging!

**Pull request merged**